File: src/main/java/org/samoxive/safetyjim/discord/commands/Softban.java
package org.samoxive.safetyjim.discord.commands;

import net.dv8tion.jda.core.EmbedBuilder;
import net.dv8tion.jda.core.JDA;
import net.dv8tion.jda.core.Permission;
import net.dv8tion.jda.core.entities.*;
import net.dv8tion.jda.core.events.message.guild.GuildMessageReceivedEvent;
import net.dv8tion.jda.core.managers.GuildController;
import org.jooq.DSLContext;
import org.samoxive.jooq.generated.Tables;
import org.samoxive.jooq.generated.tables.records.SoftbanlistRecord;
import org.samoxive.safetyjim.discord.Command;
import org.samoxive.safetyjim.discord.DiscordBot;
import org.samoxive.safetyjim.discord.DiscordUtils;
import org.samoxive.safetyjim.discord.TextUtils;

import java.awt.*;
import java.util.Date;
import java.util.Scanner;

public class Softban extends Command {
    private String[] usages = { "softban @user [reason] | [messages to delete (days)] - softbans the user with the specified args." };

    @Override
    public String[] getUsages() {
        return usages;
    }

    @Override
    public boolean run(DiscordBot bot, GuildMessageReceivedEvent event, String args) {
        Scanner messageIterator = new Scanner(args);
        JDA shard = event.getJDA();

        Member member = event.getMember();
        User user = event.getAuthor();
        Message message = event.getMessage();
        TextChannel channel = event.getChannel();
        Guild guild = event.getGuild();
        Member selfMember = guild.getSelfMember();

        if (!member.hasPermission(Permission.BAN_MEMBERS)) {
            DiscordUtils.failMessage(bot, message, "You don't have enough permissions to execute this command! Required permission: Ban Members");
            return false;
        }

        if (!messageIterator.hasNext(DiscordUtils.USER_MENTION_PATTERN)) {
            return true;
        } else {
            // advance the scanner one step to get rid of user mention
            messageIterator.next();
        }

        User softbanUser = message.getMentionedUsers().get(0);
        Member softbanMember = guild.getMember(softbanUser);
        GuildController controller = guild.getController();

        if (!selfMember.hasPermission(Permission.BAN_MEMBERS)) {
            DiscordUtils.failMessage(bot, message, "I don't have enough permissions to do that!");
            return false;
        }

        if (user.getId().equals(softbanUser.getId())) {
            DiscordUtils.failMessage(bot, message, "You can't softban yourself, dummy!");
            return false;
        }

        if (!DiscordUtils.isBannable(softbanMember, selfMember)) {
            DiscordUtils.failMessage(bot, message, "I don't have enough permissions to do that!");
            return false;
        }

        String arguments = TextUtils.seekScannerToEnd(messageIterator);
        String[] argumentsSplit = arguments.split("\\|");
        String reason = argumentsSplit[0];
        reason = reason.equals("") ? "No reason specified" : reason.trim();
        String timeArgument = null;

        if (argumentsSplit.length > 1) {
            timeArgument = argumentsSplit[1];
        }

        int days;

        if (timeArgument != null) {
            try {
                days = Integer.parseInt(timeArgument.trim());
            } catch (NumberFormatException e) {
                DiscordUtils.failMessage(bot, message, "Invalid day count, please try again.");
                return false;
            }
        } else {
            days = 1;
        }

        if (days < 1 || days > 7) {
            DiscordUtils.failMessage(bot, message, "The amount of days must be between 1 and 7.");
            return false;
        }

        Date now = new Date();

        EmbedBuilder embed = new EmbedBuilder();
        embed.setTitle("Softbanned from " + guild.getName());
        embed.setColor(new Color(0x4286F4));
        embed.setDescription("You were softbanned from " + guild.getName());
        embed.addField("Reason:", TextUtils.truncateForEmbed(reason), false);
        embed.setFooter("Softbanned by " + DiscordUtils.getUserTagAndId(user), null);
        embed.setTimestamp(now.toInstant());

        DiscordUtils.sendDM(softbanUser, embed.build());

        try {
            String auditLogReason = String.format("Softbanned by %s - %s", DiscordUtils.getUserTagAndId(user), reason);
            controller.ban(softbanMember, days, auditLogReason).complete();
            controller.unban(softbanUser).complete();

            DSLContext database = bot.getDatabase();

            SoftbanlistRecord record = database.insertInto(Tables.SOFTBANLIST,
                                                           Tables.SOFTBANLIST.USERID,
                                                           Tables.SOFTBANLIST.MODERATORUSERID,
                                                           Tables.SOFTBANLIST.GUILDID,
                                                           Tables.SOFTBANLIST.SOFTBANTIME,
                                                           Tables.SOFTBANLIST.DELETEDAYS,
                                                           Tables.SOFTBANLIST.REASON)
                                               .values(softbanUser.getId(),
                                                       user.getId(),
                                                       guild.getId(),
                                                       now.getTime() / 1000,
                                                       days,
                                                       reason)
                                               .returning(Tables.SOFTBANLIST.ID)
                                               .fetchOne();

            DiscordUtils.createModLogEntry(bot, shard, message, softbanMember, reason, "softban", record.getId(), null, false);
            DiscordUtils.sendMessage(channel, "Softbanned " + DiscordUtils.getUserTagAndId(softbanUser));
        } catch (Exception e) {
            DiscordUtils.failMessage(bot, message, "Could not softban the specified user. Do I have enough permissions?");
        }

        return false;
    }
}

File: src/main/java/org/samoxive/safetyjim/discord/commands/Softban.kt
package org.samoxive.safetyjim.discord.commands

import java.awt.Color
import java.util.*
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.TextChannel
import net.dv8tion.jda.api.entities.User
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent
import org.samoxive.safetyjim.database.SettingsEntity
import org.samoxive.safetyjim.database.SoftbanEntity
import org.samoxive.safetyjim.database.SoftbansTable
import org.samoxive.safetyjim.discord.*

private const val ACTION_REASON = "Softban threshold exceeded."

suspend fun softbanAction(guild: Guild, channel: TextChannel?, settings: SettingsEntity, modUser: User, softbanUser: User, reason: String, callDepth: Int = 0) {
    val now = Date()

    val embed = EmbedBuilder()
    embed.setTitle("Softbanned from ${guild.name}")
    embed.setColor(Color(0x4286F4))
    embed.setDescription("You were softbanned from ${guild.name}")
    embed.addField("Reason:", truncateForEmbed(reason), false)
    embed.setFooter("Softbanned by ${modUser.getUserTagAndId()}", null)
    embed.setTimestamp(now.toInstant())

    softbanUser.trySendMessage(embed.build())
    val auditLogReason = "Softbanned by ${modUser.getUserTagAndId()} - $reason"
    guild.ban(softbanUser, 1, auditLogReason).await()
    guild.unban(softbanUser).await()

    val record = SoftbansTable.insertSoftban(
        SoftbanEntity(
            userId = softbanUser.idLong,
            moderatorUserId = modUser.idLong,
            guildId = guild.idLong,
            softbanTime = now.time / 1000,
            reason = reason,
            pardoned = false
        )
    )

    createModLogEntry(guild, channel, settings, modUser, softbanUser, reason, ModLogAction.Softban, record.id)

    if (settings.softbanThreshold != 0) {
        val softbanCount = SoftbansTable.fetchUserActionableSoftbanCount(guild, softbanUser)
        if (softbanCount >= settings.softbanThreshold) {
            val expirationDate = settings.getSoftbanActionExpirationDate()
            executeModAction(settings.softbanAction, guild, channel, settings, modUser, softbanUser, ACTION_REASON, expirationDate, callDepth)
        }
    }
}

class Softban : Command() {
    override val usages = arrayOf("softban @user [reason] - soft bans the user with the specified arguments.")

    override suspend fun run(bot: DiscordBot, event: GuildMessageReceivedEvent, settings: SettingsEntity, args: String): Boolean {
        val messageIterator = Scanner(args)

        val member = event.member!!
        val user = event.author
        val message = event.message
        val channel = event.channel
        val guild = event.guild
        val selfMember = guild.selfMember

        if (!member.hasPermission(Permission.BAN_MEMBERS)) {
            message.failMessage("You don't have enough permissions to execute this command! Required permission: Ban Members")
            return false
        }

        if (args.isEmpty()) {
            return true
        }

        val (searchResult, softbanUser) = messageIterator.findUser(message, isForBan = true)
        if (searchResult == SearchUserResult.NOT_FOUND || (softbanUser == null)) {
            message.failMessage("Could not find the user to softban!")
            return false
        }

        if (searchResult == SearchUserResult.GUESSED) {
            message.askConfirmation(bot, softbanUser) ?: return false
        }

        val softbanMember = guild.fetchMember(softbanUser)

        if (!selfMember.hasPermission(Permission.BAN_MEMBERS)) {
            message.failMessage("I don't have enough permissions to do that!")
            return false
        }

        if (user == softbanUser) {
            message.failMessage("You can't softban yourself, dummy!")
            return false
        }

        if (softbanMember != null && !softbanMember.isBannableBy(selfMember)) {
            message.failMessage("I don't have enough permissions to do that!")
            return false
        }

        var reason = messageIterator.seekToEnd()
        reason = if (reason == "") "No reason specified" else reason

        try {
            softbanAction(guild, channel, settings, user, softbanUser, reason)
            message.successReact()
            channel.sendModActionConfirmationMessage(settings, "Softbanned ${softbanUser.getUserTagAndId()}")
        } catch (e: Exception) {
            message.failMessage("Could not softban the specified user. Do I have enough permissions?")
        }

        return false
    }
}
